feat: Separate Knowledge system + Multi-Agent RAG Optimization

azep-ninja opened the pull request:
# Relates to:


# Risks

Low Risk, as the new RAG system is completely separate, and must be enabled by the user to use as of now.

# Background

## What does this PR do?

## What kind of change is this?

Improvements (misc. changes to existing features)

This improves the knowledge system by separating it from the memories table, while allowing users to better manager their knowledge and have true separation of actions.

Additionally, allowing multiple agents to "share" knowledge to avoid duplicating entries for better optimizations.

All database adapters were updated to have the new functions to create/get/search/delete knowledge.

## Why are we doing this? Any context or related work?

This change is a critical change as it helps the agent scale, reduce database overlap, and separates the agents actions, to ensure proper segregation.

Additionally, this also allows users to create a knowledge folder of txt/md files and reference them in their character knowledge, without having to paste the full contents in character file.

# Documentation changes needed?

My changes require a change to the project documentation.

Minimal changes to exiting knowledge docs, to outline how to enable the new rag knowledge system.


# Testing

## Where should a reviewer start?

## Detailed testing steps

- Enable the new rag knowledge 
![image](https://github.com/user-attachments/assets/4e1ca71f-800c-47f3-9bd4-6c4103ea2aa5)

- Create a "knowledge" folder within your characters folder
![image](https://github.com/user-attachments/assets/62b6bfd3-29ad-4a55-81fc-78fcd7d26f6e)

- Add files in the root of knowledge, or create additional subfolders for organization

- Set your knowledge items in your character file directly, or objects specifying the path, and if it should be a shared item or not, depending on multi characters or not.
![image](https://github.com/user-attachments/assets/5602960c-2aef-4085-82a9-e13624acf1b6)

- Start agent(s) as normal, and watch knowledge get processed!

## Screenshots
### Before
### After

Knowledge table with information and split into chunks, along with main document, shared documents, or agent specific documents in the database.
![image](https://github.com/user-attachments/assets/ce69f38e-7fe0-4772-9137-1f6dae424006)

# Deploy Notes
## Database changes
 - New knowledge table is created
 -

## Deployment instructions

## Discord username
ninja_dev

thisisomar commented on packages/core/src/ragknowledge.ts:
```diff
@@ -0,0 +1,336 @@
+                content: {
+                    text: item.content.text,
+                    metadata: {
+                        ...item.content.metadata,
+                        isMain: true
+                    }
+                },
+                embedding: mainEmbedding,
+                createdAt: Date.now()
+            });
+
+            // Generate and store chunks
+            const chunks = await splitChunks(processedContent, 512, 20);
+
+            for (const [index, chunk] of chunks.entries()) {
+                const chunkEmbeddingArray = await embed(this.runtime, chunk);
+                const chunkEmbedding = new Float32Array(chunkEmbeddingArray);
+                const chunkId = `${item.id}-chunk-${index}` as UUID;
+
+                await this.runtime.databaseAdapter.createKnowledge({
+                    id: chunkId,
+                    agentId: this.runtime.agentId,
+                    content: {
+                        text: chunk,
+                        metadata: {
+                            ...item.content.metadata,
+                            isChunk: true,
+                            originalId: item.id,
+                            chunkIndex: index
+                        }
+                    },
+                    embedding: chunkEmbedding,
+                    createdAt: Date.now()
+                });
+            }
+        } catch (error) {
+            elizaLogger.error(`Error processing knowledge ${item.id}:`, error);
+            throw error;
+        }
+    }
```
any reason you don't use something like this from langchain https://js.langchain.com/docs/how_to/recursive_text_splitter/ 

they handle a lot thats probably not taken into consideration

thisisomar replied:
also sorry I just noticed, we have this function exposed from the core plugin 
https://github.com/elizaOS/eliza/blob/76d4f42e5763f638095d1179b2e88afc03fb8896/packages/core/src/generation.ts#L696

you could probably use this potentially or just use it your own way. not sure the context of the generation file

azep-ninja replied:
The splitChunks, which is used, is already using langchain actually, so looks like this was already the default in place.

![image](https://github.com/user-attachments/assets/11340217-b522-4bb7-b702-f734bc0f4e23)

LinuxIsCool commented:
Trying to run this locally. Getting: 

```
 ["⛔ Error parsing character from /characters/character.character.json: Error: Character configuration validation failed:\nknowledge.0: Expected string, received object\nknowledge.1: Expected string, received object"]
 ```

azep-ninja commented:
> Error parsing character from

@LinuxIsCool , good catch. Updated the character validation now. Can you pull the latest version and try.

Updated "package/core/src/environment.ts" file within CharacterSchema, replacing "knowledge:"

With

`knowledge: z.array(
        z.union([
            z.string(),
            z.object({
                path: z.string(),
                shared: z.boolean().optional()
            })
        ])
    ).optional(),`

LinuxIsCool commented:
@azep-ninja 

I'm running it now without errors on the latest version, however, it's not working as expected. The agent is running but there is no prompt in the console. 

I am seeing: 

["✓ REST API bound to 0.0.0.0:3003. If running locally, access it at http://localhost:3003."]

However, I'm not sure how to access the REST API. 

And when I try the client with  pnpm start:client 

Then, I see my agent in the dashboard, but there is not response when I post a message, and I see in the client console: 

```
> eliza@ start:client /home/ygg/Workspace/Eliza/akemi_rag
> pnpm --dir client dev


> eliza-client@0.1.7-alpha.2 dev /home/ygg/Workspace/Eliza/akemi_rag/client
> vite


  VITE v5.4.11  ready in 158 ms

  ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
  ➜  press h + enter to show help
(node:118597) ExperimentalWarning: CommonJS module /home/ygg/Workspace/Eliza/akemi_rag/node_modules/tailwindcss/lib/lib/load-config.js is loading ES Module /home/ygg/Workspace/Eliza/akemi_rag/client/tailwind.config.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
6:35:05 PM [vite] http proxy error: /agents
AggregateError [ECONNREFUSED]:
    at internalConnectMultiple (node:net:1128:18)
    at afterConnectMultiple (node:net:1693:7)
6:35:06 PM [vite] http proxy error: /agents
AggregateError [ECONNREFUSED]:
    at internalConnectMultiple (node:net:1128:18)
    at afterConnectMultiple (node:net:1693:7) (x2)
6:35:08 PM [vite] http proxy error: /agents
AggregateError [ECONNREFUSED]:
    at internalConnectMultiple (node:net:1128:18)
    at afterConnectMultiple (node:net:1693:7) (x3)
    ```

LinuxIsCool commented:
The client is now working properly after nvm use latest and re-install and re-build. So I will now test out the rag functionality.

LinuxIsCool commented:
The RAG seems to be working! 

Although now I have added a 5MB text file and the RAG system is currently hanging after about 5 minutes. I'm hoping it finishes processing soon and is not hanging indefinitely.

LinuxIsCool commented:
I think testing with big files is a good idea. It does seem to hang. But when I restart the agent then the knowledge is loaded. Are there ways that it could be sped up?

shakkernerd commented on packages/adapter-postgres/schema.sql:
```diff
@@ -24,9 +25,6 @@ BEGIN
     -- Then check for Ollama
     ELSIF current_setting('app.use_ollama_embedding', TRUE) = 'true' THEN
         RETURN 1024;  -- Ollama mxbai-embed-large dimension
-    -- Then check for GAIANET
-    ELSIF current_setting('app.use_gaianet_embedding', TRUE) = 'true' THEN
-        RETURN 768;  -- Gaianet nomic-embed dimension
```
Resolved!

iwasfirsthere commented:
If I want to try this RAGknowledge, what I need to do?

Will it work with my cloned repo from Eliza? Not develop one. Or I need to enter develop and try it there?

iwasfirsthere commented:
I've cloned your repo, added my knowledge file under character new folder, when I run the agent - it shows that knowledge files are loaded and shows the content inside, but when I ask agent a question about something that he should know from knowledge files, he says I dont know.

augchan42 commented:
Note that only the following file formats are supported:

```
 await this.ragKnowledgeManager.processFile({
                            path: contentItem,
                            content: content,
                            type: fileExtension as 'pdf' | 'md' | 'txt',
                            isShared: isShared
                        });
```

iwasfirsthere commented:
Could you please let me know how should I add links to my knowledge file and make sure agent reads them correctly, send them when asked?

It was working, but I adjusted style of my character (did all the data deletion before running again and so) but it 50% of the time now gives random links that does not exists.

augchan42 commented:
also please note that double byte characters like Chinese are not supported, as the langchain chunking lib thats used strips them, and the preprocess function also strips them

iwasfirsthere commented:
What is the best way to run this?

If I download your repo, I keep getting errors on pglite and other plugins/adapters.

I was managed to run it before somehow on my own eliza repo, but only with downloading ragknowledge.ts file and adding some other stuff with cursor. Now I ruined it and can't run anyhow. Please help

y4my4my4m commented:
@azep-ninja can this be used with the twitter spaces? it doesnt seem to make use of RAG

LinuxIsCool commented:
What does the isShared parameter do?

augchan42 commented:
When we retrieve knowledge, we do an OR for agent specific knowledge, and and for knowledge marked as shared.  Shared knowledge is available to all agents (even if only 1 character card loads that shared knowledge)

LinuxIsCool commented:
I don't think it's ideal that all characters access shared knowledge by default. They should still have to declare the knowledge file in their character file. 

Also, I think there is a bug in this PR in which entire documents are embedded and they are being retrieved at runtime which is blowing up my prompt. To address this, I changed the sql in adapter-sqlite/index.ts in getKnowledge to be: 
```
let sql = `SELECT * FROM knowledge WHERE (isMain = 0) AND (agentId = ? OR isShared = 1)`;
```
That's here: https://github.com/elizaOS/eliza/blob/678144b8cefe6225a7107400f83f65f2be929386/packages/adapter-sqlite/src/index.ts#L768 

But I think there might be a more holistic solution to this problem, not sure what the right intervention would be. Maybe just don't embed main documents?

LinuxIsCool commented:
So there's currently an issue where I can't have multiple agents access the same knowledge, without making that knowledge accessible to all agents.

augchan42 commented:
You'll need to load the same knowledge as character specific for those
agents then.

On Thu, Jan 30, 2025, 2:42 AM Shawn Anderson ***@***.***>
wrote:

> So there's currently an issue where I can't have multiple agents access
> the same knowledge, without making that knowledge accessible to all agents.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/elizaOS/eliza/pull/1620#issuecomment-2622542667>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE32LKKZVDWSLIS3PDOFPZL2NEOIPAVCNFSM6AAAAABUN42OKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMRSGU2DENRWG4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

LinuxIsCool commented:
If I add the same knowledge on two different agents with shared=false then I get an error when running 
 loading knowledge. Maybe it still works even through it errors?

LinuxIsCool commented:
Additionally, I think that a knowledge file is currently not re-processed when it is updated.
